File: solutions/2026-05-26/cpp/2026-05-26-count-the-number-of-special-characters-i.cpp
// Problem: Count the Number of Special Characters I
// Link: https://leetcode.com/problems/count-the-number-of-special-characters-i/
//
// Approach:
// We can use two hash sets (or boolean arrays of size 26 for simplicity given the constraints)
// to keep track of lowercase and uppercase letters encountered in the string.
// We iterate through the input string `word`. For each character, we check if it's lowercase or uppercase.
// If it's lowercase, we add its corresponding character value (e.g., 'a' to 0, 'b' to 1) to the lowercase set.
// If it's uppercase, we add its corresponding character value (e.g., 'A' to 0, 'B' to 1) to the uppercase set.
// After iterating through the entire string, we iterate from 0 to 25 (representing 'a' to 'z' or 'A' to 'Z').
// For each index `i`, if both the lowercase set and the uppercase set contain the character corresponding to `i`
// (i.e., 'a' + i and 'A' + i), then this letter is special. We increment a counter for special letters.
// Finally, we return the total count of special letters.
//
// Time Complexity: O(N), where N is the length of the input string `word`.
// We iterate through the string once to populate the sets and then iterate up to 26 times (constant) to count special characters.
//
// Space Complexity: O(1), as the size of the boolean arrays used to track letters is fixed at 26,
// independent of the input string length.

#include <string>
#include <vector>
#include <cctype>

class Solution {
public:
    int numberOfSpecialChars(std::string word) {
        // Boolean arrays to track presence of lowercase and uppercase letters.
        // Index 0 corresponds to 'a'/'A', index 1 to 'b'/'B', and so on.
        std::vector<bool> lowerPresent(26, false);
        std::vector<bool> upperPresent(26, false);

        // Iterate through the input string to populate the presence arrays.
        for (char c : word) {
            if (std::islower(c)) {
                // Mark the presence of the lowercase character.
                lowerPresent[c - 'a'] = true;
            } else if (std::isupper(c)) {
                // Mark the presence of the uppercase character.
                upperPresent[c - 'A'] = true;
            }
        }

        int specialCharCount = 0;
        // Iterate through all possible letters (a-z or A-Z).
        for (int i = 0; i < 26; ++i) {
            // A letter is special if both its lowercase and uppercase forms are present.
            if (lowerPresent[i] && upperPresent[i]) {
                specialCharCount++;
            }
        }

        // Return the total count of special letters.
        return specialCharCount;
    }
};

File: solutions/2026-05-26/python/2026-05-26-count-the-number-of-special-characters-i.py
```python
# Problem Summary: Count letters that appear in both lowercase and uppercase within a string.
# Link: https://leetcode.com/problems/count-the-number-of-special-characters-i/
# Approach Explanation:
# We can use two sets to store the lowercase and uppercase letters encountered in the string.
# Iterate through the input string 'word'.
# For each character, convert it to lowercase and add it to the 'lowercase_chars' set.
# Also, convert it to uppercase and add it to the 'uppercase_chars' set.
# After processing the entire string, iterate through the 'lowercase_chars' set.
# For each lowercase character, check if its uppercase equivalent is present in the 'uppercase_chars' set.
# If both are present, it means this letter is "special", so increment a counter.
# Finally, return the total count of special letters.
# Time Complexity Analysis: O(N), where N is the length of the input string 'word'.
# We iterate through the string once to populate the sets, and then iterate through the lowercase set (at most 26 elements).
# Set insertions and lookups are O(1) on average.
# Space Complexity Analysis: O(1), as the size of the sets is bounded by the number of lowercase and uppercase English letters (26 each).
class Solution:
    def numberOfSpecialChars(self, word: str) -> int:
        # Initialize a set to store all lowercase characters found
        lowercase_chars = set()
        # Initialize a set to store all uppercase characters found
        uppercase_chars = set()
        
        # Iterate through each character in the input string
        for char in word:
            # Add the lowercase version of the character to the lowercase_chars set
            lowercase_chars.add(char.lower())
            # Add the uppercase version of the character to the uppercase_chars set
            uppercase_chars.add(char.upper())
            
        # Initialize a counter for special characters
        special_count = 0
        
        # Iterate through each unique lowercase character found
        for lower_char in lowercase_chars:
            # Check if the uppercase version of this lowercase character also exists in the uppercase_chars set
            if lower_char.upper() in uppercase_chars:
                # If both lowercase and uppercase versions are present, increment the special_count
                special_count += 1
                
        # Return the total count of special characters
        return special_count

```